### PR TITLE
[Feat] Bootstrap types & EA Input Parameter types

### PR DESCRIPTION
**Pull request by justinkaseman**
## Description

Refactor Bootstrap for no `any` types. 
Adds types to EA request input parameters.

## Changes

- Removes the internal types package `@chainlink/types`. EA shared types will now be exported from the root of `@chainlink/ea-bootstrap`. This keeps the framework types coupled closer with the framework itself. (sc-20168)
- Framework types are kept in `.ts` files to enable the TS linter to check types. (sc-15806)
- Adds missing types & type guards for no any types in `@chainlink/ea-bootstrap` (sc-19690 & sc-26500)
- Adds `limits` to Adapter Context to hold the calculated rate limits (as compared to the raw API tier rate limits)
- Adds README descriptions to each middleware and module
- Move `WSBaseURL` to its own object in `Config`
- Upgrades dependencies:
   - LRU-cache
   - sinon
   - node-redis -  `1.0.0` -> `1.0.3` to retrieve newly exposed type interfaces (sc-23811)

## Quality Assurance

- [x] Ran `yarn changeset` if adapter source code was changed
- [x] If a new adapter was made, or an existing one was modified so that its environment variables have changed, update the relevant `<ADAPTER_PACKAGE>/schemas/env.json` and `<ADAPTER_PACKAGE>/README.md`
- [x] If a new adapter was made, or an existing one was modified so that its environment variables have changed, update the relevant `infra-k8s` configuration file.
- [x] The branch naming follows git flow (`feature/x`, `chore/x`, `release/x`, `hotfix/x`, `fix/x`) or is created from Clubhouse/Shortcut
- [ ] This is related to a maximum of one Clubhouse/Shortcut story or GitHub issue
- [x] Types are safe (avoid TypeScript/TSLint features like any and disable, instead use more specific types)

### CODE REVIEW

File: packages/core/bootstrap/src/lib/middleware/cache-warmer/reducer.ts [hunk 1]
```diff
@@ -107,18 +108,20 @@ export const subscriptionsReducer = createReducer<SubscriptionState>({}, (builde
...
-              childRequestData,
-              name,
-            })
-          batchWarmer.origin[name] = [...uniqueBatchableValue]
+          const uniqueBatchableValue = new Set<typeof batchWarmer['origin']['name']>()
+          const originalValue = batchWarmer.origin[name]
+          uniqueBatchableValue.add(originalValue)
+          const incomingValue =
+            childRequestData[name] ||
+            (typeof childRequestData.data === 'object' &&
```
Comment (KuphJr) [inline packages/core/bootstrap/src/lib/middleware/cache-warmer/reducer.ts:118]: Should this be narrowed further?

File: packages/core/bootstrap/src/lib/middleware/cache-warmer/util.ts [hunk 2]
```diff
@@ -36,24 +36,20 @@ type GroupedBatches = {
...
   const batchesByPath: GroupedBatches = {}
   for (const { name, limit } of batchablePropertyPath) {
-    if (origin[name]) {
-      const batchedValues = origin[name]
-      if (limit) {
-        batchesByPath[name] = splitValuesIntoBatches(limit, batchedValues)
-      } else {
-        batchesByPath[name] = [batchedValues]
-      }
+    const batchedValues = origin[name] as any // TODO, log this
```
Comment (KuphJr) [inline packages/core/bootstrap/src/lib/middleware/cache-warmer/util.ts:43]: check this

File: packages/core/bootstrap/src/lib/middleware/ws/epics.ts [hunk 3]
```diff
@@ -188,15 +189,13 @@ export const connectEpic: Epic<AnyAction, AnyAction, { ws: RootState }, any> = (
...
-      const WebSocketCtor = WebSocketClassProvider.get()
       const wsSubject = webSocket({
         url,
-        protocol, // TODO: Double check this
+        protocol,
         deserializer,
         serializer,
         openObserver,
         closeObserver,
-        WebSocketCtor: WebSocketCtor as any, // TODO: fix types don't match
```
Comment (KuphJr) [inline packages/core/bootstrap/src/lib/middleware/ws/epics.ts:199]: Ask @alejoberardino

File: packages/core/bootstrap/src/lib/modules/error.ts [hunk 4]
```diff
@@ -1,14 +1,14 @@
...
+import type { AdapterErrorResponse } from '../../types'
 
 export class AdapterError extends Error {
   jobRunID: string
   status: string
   statusCode: number
   name: string
   message: string
-  cause: any
+  cause: unknown
```
Comment (KuphJr) [inline packages/core/bootstrap/src/lib/modules/error.ts:9]: Could we extend this from nodes error type

File: packages/core/bootstrap/src/lib/modules/requester.ts [hunk 5]
```diff
@@ -32,13 +33,6 @@ export class Requester {
       config.timeout = !isNaN(timeout) ? timeout : 3000
     }
 
-    if (!customError) customError = getFalse
-    if (typeof customError !== 'function') {
-      delay = retries
-      retries = customError
-      customError = getFalse
-    }
-
```
Comment (KuphJr) [inline packages/core/bootstrap/src/lib/modules/requester.ts:41]: let's test this. delay = retries???

File: packages/core/bootstrap/src/lib/modules/validator.ts [hunk 6]
```diff
@@ -119,25 +129,29 @@ export class Validator {
...
-        this.validated.overrides.get(adapter.toLowerCase())?.get(defaultSymbol.toLowerCase()) ||
-        defaultSymbol
-      )
-    const multiple: string[] = []
-    for (const sym of defaultSymbol) {
-      const overrided = this.validated.overrides.get(adapter.toLowerCase())?.get(sym.toLowerCase())
-      if (!overrided) multiple.push(sym)
-      else multiple.push(overrided)
+  overrideSymbol = (adapter: string, symbol = this.validated.data.base): NestableValue => {
+    if (symbol === undefined) throw this.throwInvalid(`Required parameter not supplied: base`)
```
Comment (KuphJr) [inline packages/core/bootstrap/src/lib/modules/validator.ts:133]: This error message probably shouldn't say base

File: packages/core/bootstrap/src/lib/metrics/util.ts [hunk 7]
```diff
@@ -80,8 +80,20 @@ export const getFeedId = (input: AdapterRequest): string => {
        * With batched requests, the base can either be an array of bases, or a single base.
        * The same type constraints apply to the quote param.
        */
-      if (base) {
-        return `${buildSymbolString(base)}` + (quote ? `/${buildSymbolString(quote)}` : '')
+      if (
+        base &&
```
[the remark this answers is not in the document]
Comment (boxhock) [reply packages/core/bootstrap/src/lib/metrics/util.ts:84]: This is marked as resolved, but looks like we still have `base &&` here. Is this intentional?

**Review by boxhock — DISMISSED** [review]

**Review by boxhock — APPROVED** [review]

**Pull request merged**